docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+
+### API
+
+`ETCDOperatorTask` is the new custom resource that will be introduced. This API will be in `v1alpha1` version and will be subject to change. We will be respecting [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
+
+```yaml
+apiVersion: druid.gardener.cloud/v1alpha1
+kind: EtcdOperatorTask
+metadata:
+    name: <name of operator task resource>
+    namespace: <cluster namespace>
+    generation: <specific generation of the desired state>
+spec:
+    taskType: <type/category of supported out-of-band task>
+    ttlSecondsAfterFinished: <time-to-live to garbage collect the custom resource>
+    taskConfig: <task specific configuration>
+status:
+    observedGeneration: <specific observedGeneration of the resource>
+    taskStatus: < overall status of the task >
+    initiatedAt: <time of intiation of this operation>
+    lastErrors:
+      - code: <error-code>
+        description: <description of the error>
+        lastUpdateTime: <time the error was reported>
+    lastOperation:
+      name: <operation-name>
+      state: <task state as seen at the completion of last operation>
+      lastTransitionTime: <time of transition to this state>
+      reason: <reason/message if any>
+    conditions:
+      - type: <type of condition>
+        status: <status of the condition, one of True, False, Unknown>
+        lastTransitionTime: <last time the condition transitioned from one status to another>
+        reason: <programmatic identifier indicating the reason for the condition's last transition>
+        message: <human readable message indicating details about the transition>
+```
+
+> NOTE: The above custom resource YAML serves as a template which will be used to further create a CRD and Golang APIs.
+
+### Golang API
+
+```go
+// EtcdOperatorTask represents an out-of-band operator task.
+type EtcdOperatorTask struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // Spec is the specification of the task.
+  Spec EtcdOperatorTaskSpec
+  // Status is most recently observed status of the task.
+  Status EtcdOperatorTaskStatus
+}
+```
+
+#### Spec
+
+The authors propose that the following fields should be specified in the spec (desired state) of the `EtcdOperatorTask` custom resource.
+
+* To capture the type of `out-of-band` operator task to be performed, `taskType` field should be defined in spec. It can have values of all supported "out-of-band" operations/tasks eg. "OnDemandSnaphotTask", "QuorumLossRecoveryTask" etc.
+* To capture the configuration specific to each task, a `.spec.taskConfig` field is defined in spec as type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102) as each task may have different specifications.
```

Comment:
```suggestion
* To capture the configuration specific to each task, a `.spec.Config` field is defined in spec as type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102) as each task may have different specifications.
```

Reply:
We don't do the same change request in line 136 for`taskType` as we do here for `taskConfig`. Its a litt confusing to read what is the agreement on semantics. 
```suggestion
* To capture the configuration specific to each task, a `taskConfig` field should be defined in spec as type [RawExtension](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L49-L102) as each task may have different specifications.
```